### PR TITLE
disttask: maintain managed nodes separately

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #49008

Problem Summary:

### What changed and how does it work?
- `GetEligibleInstances` returns target node that task can run, if no node returned, we use all.
- add nodeManager to maintain managed node

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/backfilling_dist_scheduler_test.go [hunk 1]
```diff
@@ -173,8 +172,7 @@ func TestBackfillingSchedulerGlobalSortMode(t *testing.T) {
 	taskID, err := mgr.CreateTask(ctx, task.Key, proto.Backfill, 1, task.Meta)
 	require.NoError(t, err)
 	task.ID = taskID
-	serverInfos, _, err := sch.GetEligibleInstances(context.Background(), task)
-	require.NoError(t, err)
+	serverInfos := []string{":4000"}
```
Comment (ywqzzy) [inline pkg/ddl/backfilling_dist_scheduler_test.go:175]:
```suggestion
	execIDs := []string{":4000"}
```

File: pkg/disttask/framework/scheduler/interface.go [hunk 2]
```diff
@@ -105,8 +104,10 @@ type Extension interface {
 
 	// GetEligibleInstances is used to get the eligible instances for the task.
 	// on certain condition we may want to use some instances to do the task, such as instances with more disk.
-	// The bool return value indicates whether filter instances by role.
-	GetEligibleInstances(ctx context.Context, task *proto.Task) ([]*infosync.ServerInfo, bool, error)
+	// if returned instances is empty, it means all instances are eligible.
+	// TODO: run import from server disk using framework makes this logic complicated,
+	// the instance might not be managed by framework.
+	GetEligibleInstances(ctx context.Context, task *proto.Task) ([]string, error)
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/interface.go:110]: Maybe we can remove this interface?

File: pkg/disttask/framework/scheduler/scheduler_manager_test.go [hunk 3]
```diff
@@ -39,13 +39,13 @@ func TestCleanUpRoutine(t *testing.T) {
 	defer ctrl.Finish()
 	ctx := context.Background()
 	ctx = util.WithInternalSourceType(ctx, "scheduler_manager")
-	mockCleanupRountine := mock.NewMockCleanUpRoutine(ctrl)
+	mockCleanupRoutine := mock.NewMockCleanUpRoutine(ctrl)
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/scheduler_manager_test.go:42]:
```suggestion
	mockCleanupTask := mock.NewMockCleanUpTask(ctrl)
```

File: pkg/disttask/framework/scheduler/scheduler.go [hunk 4]
```diff
@@ -624,26 +569,38 @@ func (s *BaseScheduler) onNextStage() (err error) {
...
+// getEligibleNodes returns the eligible(live) nodes for the task.
+// if the task can only be scheduled to some specific nodes, return them directly,
+// we don't care liveliness of them.
+func (s *BaseScheduler) getEligibleNodes() ([]string, error) {
+	serverNodes, err := s.GetEligibleInstances(s.ctx, s.Task)
+	if err != nil {
+		return nil, err
+	}
+	logutil.Logger(s.logCtx).Debug("eligible instances", zap.Int("num", len(serverNodes)))
+	if len(serverNodes) == 0 {
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/scheduler.go:581]: this part is same as before
if task can only run on some nodes(len > 0), we use it, else we can use all managed nodes

File: pkg/disttask/framework/scheduler/main_test.go [hunk 5]
```diff
@@ -29,24 +30,36 @@ type SchedulerManagerForTest interface {
...
-func (dm *Manager) DelRunningTask(id int64) {
-	dm.delScheduler(id)
+func (sm *Manager) DelRunningTask(id int64) {
+	sm.delScheduler(id)
 }
 
 // DoCleanUpRoutine implements Scheduler.DoCleanUpRoutine interface.
-func (dm *Manager) DoCleanUpRoutine() {
-	dm.doCleanUpRoutine()
+func (sm *Manager) DoCleanUpRoutine() {
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/main_test.go:43]: ditto

File: pkg/disttask/framework/scheduler/scheduler_manager.go [hunk 6]
```diff
@@ -120,9 +121,18 @@ func (sm *Manager) Start() {
 	failpoint.Inject("disableSchedulerManager", func() {
 		failpoint.Return()
 	})
+	// init cached managed nodes
+	sm.nodeMgr.refreshManagedNodes(sm.ctx, sm.taskMgr)
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/scheduler_manager.go:125]: what if refresh failed?

File: pkg/disttask/framework/scheduler/scheduler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ywqzzy) [reply pkg/disttask/framework/scheduler/scheduler.go:581]: I get it, instance is for local mode.

File: pkg/disttask/framework/scheduler/interface.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/interface.go:110]: not now, server disk import requires this

File: pkg/disttask/framework/scheduler/scheduler_manager.go (hunk 6 above)
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/scheduler_manager.go:125]: it's called periodicly per second, here just init it to make test pass

File: pkg/disttask/framework/scheduler/nodes.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	// liveNodesCheckInterval is the tick interval of fetching all server infos from etcs.
+	nodesCheckInterval = 2 * checkTaskFinishedInterval
+)
+
+// NodeManager maintains live TiDB nodes in the cluster, and maintains the nodes
+// managed by the framework.
+type NodeManager struct {
+	// prevLiveNodes is used to record the live nodes in last checking.
+	prevLiveNodes map[string]struct{}
+	managedNodes  atomic.Pointer[[]string]
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/nodes.go:38]: add some comments for managedNodes.

File: pkg/disttask/framework/scheduler/nodes.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			nm.maintainLiveNodes(ctx, taskMgr)
+		}
+	}
+}
+
+// maintainLiveNodes manages live node info in dist_framework_meta table
+// see recoverMetaLoop in task executor for when node is insert into dist_framework_meta.
```
Comment (ywqzzy) [inline pkg/disttask/framework/scheduler/nodes.go:66]:
```suggestion
// see recoverMetaLoop in task executor for when node is inserted into dist_framework_meta.
```

**Review by ywqzzy — APPROVED** [review]
LGTM

**Review by okJiang — COMMENTED** [review]
17/28

File: pkg/disttask/framework/scheduler/main_test.go [hunk 9]
```diff
@@ -29,24 +30,36 @@ type SchedulerManagerForTest interface {
...
 
+func (s *BaseScheduler) DoBalanceSubtasks(eligibleNodes []string) error {
+	return s.doBalanceSubtasks(eligibleNodes)
+}
+
+func NewNodeManager() *NodeManager {
+	return newNodeManager()
+}
+
+func (nm *NodeManager) RefreshManagedNodes(ctx context.Context, taskMgr TaskManager) {
```
Comment (okJiang) [inline pkg/disttask/framework/scheduler/main_test.go:59]: not used

File: pkg/disttask/framework/scheduler/main_test.go [hunk 10]
```diff
@@ -29,24 +30,36 @@ type SchedulerManagerForTest interface {
...
 
 func (s *BaseScheduler) OnNextStage() (err error) {
 	return s.onNextStage()
 }
 
+func (s *BaseScheduler) DoBalanceSubtasks(eligibleNodes []string) error {
+	return s.doBalanceSubtasks(eligibleNodes)
+}
+
+func NewNodeManager() *NodeManager {
```
Comment (okJiang) [inline pkg/disttask/framework/scheduler/main_test.go:55]: why put it here?

Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/main_test.go:55]: used in cross-pkg test

File: pkg/disttask/framework/scheduler/nodes_test.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+	// fail on clean
+	mockTaskMgr.EXPECT().GetAllNodes(gomock.Any()).Return([]string{":4000", ":4001", ":4002"}, nil)
+	mockTaskMgr.EXPECT().DeleteDeadNodes(gomock.Any(), gomock.Any()).Return(errors.New("mock error"))
+	nodeMgr.maintainLiveNodes(ctx, mockTaskMgr)
+	require.Equal(t, map[string]struct{}{":4000": {}}, nodeMgr.prevLiveNodes)
+	require.True(t, ctrl.Satisfied())
+	// remove 1 node
+	mockTaskMgr.EXPECT().GetAllNodes(gomock.Any()).Return([]string{":4000", ":4001", ":4002"}, nil)
+	mockTaskMgr.EXPECT().DeleteDeadNodes(gomock.Any(), gomock.Any()).Return(nil)
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/nodes_test.go:73]: not needed, only 1 callsite

File: pkg/disttask/framework/scheduler/scheduler.go [hunk 12]
```diff
@@ -599,17 +551,10 @@ func (s *BaseScheduler) onNextStage() (err error) {
 		}
 	}
 
-	serverNodes, filter, err := s.GetEligibleInstances(s.ctx, s.Task)
+	serverNodes, err := s.getEligibleNodes()
 	if err != nil {
 		return err
 	}
-	logutil.Logger(s.logCtx).Debug("eligible instances", zap.Int("num", len(serverNodes)))
-	if filter {
```
Comment (okJiang) [inline pkg/disttask/framework/scheduler/scheduler.go:607]: What will we do here when existing some dead nodes? Depend on the next `balance` to remove the dead nodes?

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: pkg/disttask/framework/scheduler/scheduler.go (hunk 12 above)
Comment (D3Hunter) [reply pkg/disttask/framework/scheduler/scheduler.go:607]: yes

**Review by okJiang — APPROVED** [review]

**Pull request merged**